libraries/InternalFileSytem/src/flash/flash_nrf5x.c

```diff
@@ -44,11 +45,52 @@ extern uint32_t __flash_arduino_start[];
 // MACRO TYPEDEF CONSTANT ENUM DECLARATION
 //--------------------------------------------------------------------+
 static SemaphoreHandle_t _sem = NULL;
+static uint32_t _flash_op_result = NRF_EVT_FLASH_OPERATION_SUCCESS;
 
 void flash_nrf5x_event_cb (uint32_t event)
 {
-//  if (event != NRF_EVT_FLASH_OPERATION_SUCCESS) LOG_LV1("IFLASH", "Flash op Failed");
-  if ( _sem ) xSemaphoreGive(_sem);
+  if ( _sem ) {
+    // Record the result, for consumption by fal_erase or fal_program
+    // Used to reattempt failed operations
+    _flash_op_result = event;
+
+    // Signal to fal_erase or fal_program that our async flash op is now complete
+    xSemaphoreGive(_sem);
+  } 
+}
+
+// How many retry attempts when performing flash operations 
+#define MAX_RETRY 20
+
+// When soft device is enabled, flash ops are async
+// Eventual success is reported via callback, which we await
+static uint32_t wait_for_async_flash_op_completion(uint32_t initial_result) {
+  // If initial result not NRF_SUCCESS, no need to await callback
+  // We will pass the initial result (failure) straight through
+  int32_t result = initial_result;
+
+  // Operation was queued successfully
+  if (initial_result == NRF_SUCCESS) {
+
+    // Wait for result via callback 
+    xSemaphoreTake(_sem, portMAX_DELAY);
+    
+    // If completed successfully
+    if (_flash_op_result == NRF_EVT_FLASH_OPERATION_SUCCESS) 
```

Comment:
Always use the curly braces.  This is particularly important in codebases that heavily use macros.  Small but important change.